include/zephyr/drivers/mspi.h

```diff
@@ -0,0 +1,791 @@
+/*
+ * Copyright (c) 2024, Ambiq Micro Inc. <www.ambiq.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public APIs for MSPI driver
+ */
+
+#ifndef ZEPHYR_INCLUDE_MSPI_H_
+#define ZEPHYR_INCLUDE_MSPI_H_
+
+#include <errno.h>
+
+#include <zephyr/sys/__assert.h>
+#include <zephyr/types.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief MSPI Driver APIs
+ * @defgroup mspi_interface MSPI Driver APIs
+ * @ingroup io_interfaces
+ * @{
+ */
+
+/**
+ * @brief MSPI operational mode
+ */
+enum mspi_op_mode {
+	MSPI_OP_MODE_MASTER         = 0,
```

Comment:
I'm confused when an MSPI device would operate in slave mode. Maybe I'm not understanding the design of this API correctly, but it seems like devices using the MSPI API would then implement other driver classes. For example,

- flash chips would implement the flash API using the MSPI API
- PSRAM chips would configure memory mapped access during init, and be implemented as MEMC drivers (no user facing API, although perhaps one is needed to get the base address of the MEMC device)
- other devices (displays, cameras, etc) would implement their relevant driver class

Reply:
Yes, your are correct. For slave mode, I do need someone to fill the void but I'd envision it also uses the proposed the API since the controller essentially act as a device itself and the devicetree structure still stands. The `mspi_config` would govern hardware specific and platform specific hardware settings. The `mspi_dev_config` would govern the device specific hardware settings (with respect to master).

For incomming transactions from master, it is essential to register a callback using mspi_register_callback so that the upper layer may be notified.

Reply:
Ok, so slave mode would be for a case where the MCU is acting as a receiver of QSPI/OSPI commands from another device? I think I understand the use case a bit better now- are you aware of any hardware that actually would use this? I suppose some FPGAs might present themselves as a slave on a QSPI bus, I'm just not clear what the software interface would look like for a peripheral that supports acting as a QSPI slave

Reply:
Like Communication between MCU and FPGA or between two MCUs. I think these are quite flexible in terms of settings I defined in the header file and would be for specific purposes not something you could just buy from market.